source/tests/system/nidmm_driver_api_tests.cpp

```diff
@@ -0,0 +1,282 @@
+        {
+            ::grpc::ServerBuilder builder;
+            session_repository_ = std::make_unique<nidevice_grpc::SessionRepository>();
+            nidmm_library_ = std::make_unique<dmm::NiDMMLibrary>();
+            nidmm_service_ = std::make_unique<dmm::NiDMMService>(nidmm_library_.get(), session_repository_.get());
+            builder.RegisterService(nidmm_service_.get());
+
+            server_ = builder.BuildAndStart();
+            ResetStub();
+        }
+
+        virtual ~NiDMMDriverApiTest() {}
+
+        void SetUp() override
+        {
+            initialize_driver_session();
+        }
+
+        void TearDown() override
+        {
+            close_driver_session();
+        }
+
+        void ResetStub()
+        {
+            channel_ = server_->InProcessChannel(::grpc::ChannelArguments());
+            nidmm_stub_ = dmm::NiDMM::NewStub(channel_);
+        }
+
+        std::unique_ptr<dmm::NiDMM::Stub>& GetStub()
+        {
+            return nidmm_stub_;
+        }
+
+        int GetSessionId()
+        {
+            return driver_session_->id();
+        }
+
+        void expect_api_success(int error_status)
+        {
+            EXPECT_EQ(kDMMDriverApiSuccess, error_status) << get_error_message(error_status);
+        }
+
+        void initialize_driver_session()
+        {
+            ::grpc::ClientContext context;
+            dmm::InitWithOptionsRequest request;
+            request.set_resource_name("SimulatedDMM");
+            request.set_option_string("Simulate=1, DriverSetup=Model:4065; BoardType:PCI");
+            request.set_session_name("");
+            request.set_reset_device(false);
+            request.set_id_query(false);
+            dmm::InitWithOptionsResponse response;
+
+            ::grpc::Status status = GetStub()->InitWithOptions(&context, request, &response);
+            driver_session_ = std::make_unique<nidevice_grpc::Session>(response.vi());
+
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
+        }
+
+        void close_driver_session()
+        {
+            ::grpc::ClientContext context;
+            dmm::CloseRequest request;
+            request.mutable_vi()->set_id(driver_session_->id());
+            dmm::CloseResponse response;
+
+            ::grpc::Status status = GetStub()->Close(&context, request, &response);
+
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
+        }
+
+        std::string get_error_message(int error_status)
+        {
+            ::grpc::ClientContext context;
+            dmm::GetErrorMessageRequest request;
+            request.mutable_vi()->set_id(GetSessionId());
+            request.set_error_code(error_status);
+            dmm::GetErrorMessageResponse response;
+
+            ::grpc::Status status = GetStub()->GetErrorMessage(&context, request, &response);
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
```

Comment:
I think I would just have `EXPECT_EQ(kDMMDriverApiSuccess, response.status())` directly here instead of `expect_api_success(response.status())`. Just in case the call to `GetErrorMessage` fails we should avoid an infinite recursion.

Reply:
changed this